### ISSUE TITLE
[Developer] Discrepancy of width/padding of touch-layouts between Developer and KeymanWeb

### ISSUE THREAD

**Issue by eddieantonio** [keymanapp/keyman#1890]
**Describe the bug**

tl;dr: display of width/padding can look correct in Keyman Developer, but look wrong when rendered in KeymanWeb.

---

Suppose I define a row of keys in Keyman Developer for a touch-optimized layout. I let the width and padding of all of the keys set to the defaults (i.e., unspecified), which are (as of this writing), 100 and 5, respectively. Thus, the default _total_ width of a row with 8 keys with default width and padding is:

    total row width = n * default key width + n * default padding
    total row width = 8 * 100 + 8 * 5
    total row width = 840

I create another row with 7 keys: 6 default keys, and one key taking the space of two default keys. This is so **the keys can be displayed in an orthogonal grid**. This double-sized key's width can be calculated as follows:

    my key width = k * default key width + (k - 1) * default padding
    my key width = 2 * 100 + (2 - 1) * 5
    my key width = 205

Thus the total row occupies this much space:

    total row width = (n * default key width + n * default padding) + (my key width + default padding)
    total row width = (6 * 100 + 6 * 5) + (205 + 5)
    total row width = 840

Excellent! The rows have the same width! And the layout seems to be arranged as a perfect grid in Keyman Developer:

![A touch-layout with all row widths == 840 in Keyman Developer](https://user-images.githubusercontent.com/2294397/61809952-80c06600-adfb-11e9-9e0d-e2520152ed52.PNG)

I even added a row with 4 regular keys and one key the size of 4 keys, and it still looks right.

Now let's view this touch-layout on mobile:

![An incorrectly spaced layout on KeymanWeb](https://user-images.githubusercontent.com/2294397/61808879-85841a80-adf9-11e9-9dfa-0dbc1985a29d.png)

Yikes! The spacing of my two special rows has shifted. Now my **large keys occupy less width than expected** when rendered in KeymanWeb. Let's mark the discrepancies:

![kmw-render-annotated](https://user-images.githubusercontent.com/2294397/61809083-dc89ef80-adf9-11e9-8c68-e40c69a22a69.png)

**To Reproduce**
Using the attached touch-layout: 
[nrc_crk_cans.keyman-touch-layout.txt](https://github.com/keymanapp/keyman/files/3427750/nrc_crk_cans.keyman-touch-layout.txt)

 1. Load the touch layout in Keyman Developer; notice that the keys are aligned in a grid.
 2. Now, compile the keyboard, and start the local test server. Navigate to the test server with your smartphone. The layout will have changed!

**Expected behavior**

**KeymanWeb should render the keyboard as it is rendered in Keyman Developer.** Namely, it should render the given layout as a grid:

![The touch-layout, correctly rendered in Keyman Developer (duplicate)](https://user-images.githubusercontent.com/2294397/61808615-fe36a700-adf8-11e9-9db9-5f014927b6be.png)

**iPhone/iPad/Android (please complete the following information):**
 - Device: OnePlus 3T
 - OS: Android 9
 - Keyman version: Bug appears in KeymanWeb 11, and is still present in KeymanWeb 12-alpha
 - Target application: Keyman

**KeymanWeb (please complete the following information):**
 - OS:  Windows 7, macOS 10.14 Mojave, Android 9.0
 - Browser: Chrome, Firefox
 - KeymanWeb Version: 11.x~12.0
 - URL: (keyboard testing on localhost)
 
**Keyman Developer (please complete the following information):**
 - OS: Windows 7
 - Keyman Developer version: 11
 
---

Please find attached the offending touch layout!

[nrc_crk_cans.keyman-touch-layout.txt](https://github.com/keymanapp/keyman/files/3427750/nrc_crk_cans.keyman-touch-layout.txt)

---

According to @mcdurdin on Slack:

> I've noticed this before; I think the issue relates to how KeymanWeb does its rescaling of keyboard width to screen width. Your keyman-touch-layout spec looks correct.

**Comment by mcdurdin** [comment]
Great report :) I'm going to use this as an example of how to write a bug report. It's so good I don't want to fix the bug, because I want it to stay visible in the list of open bugs just so I can go back and re-read it periodically.

**Comment by mcdurdin** [comment]
This arises because of a discrepancy in the default padding values for keys. In Keyman Developer, the value is assumed to be 5 units, but in KeymanWeb, the value is 15 units.

A secondary issue is that (unnecessary) rounding in KeymanWeb leads to a more uneven looking keyboard. 

Setting the default in KeymanWeb to 5 units results in a keyboard with correctly aligned keys, but padding is insufficient.

Here's the initial image on my test device (iPhone 6 Plus, 414px), 15 unit padding:
![image](https://user-images.githubusercontent.com/4498365/66164531-3ae31200-e676-11e9-9132-abd25d54f8b1.png)

And with 5 unit padding:

![image](https://user-images.githubusercontent.com/4498365/66165107-5b5f9c00-e677-11e9-8678-ed660db8b747.png)

The outcome with 5 unit padding is definitely sub-optimal. This is not a change I am willing to make at this point in the beta -- it has a visible impact on all keyboards.

The best interim solution may be to add custom padding (`"pad":"15"`) to each key in the keyboard, which results in:

![image](https://user-images.githubusercontent.com/4498365/66165513-3a4b7b00-e678-11e9-9ec5-2108cd604c9f.png)

**Comment by mcdurdin** [comment]
On my iPhone 7, native device pixels rather than scaled through emulation in Chrome:

![image](https://user-images.githubusercontent.com/4498365/66166012-6b787b00-e679-11e9-94f6-42de81307f2c.png)

**Comment by mcdurdin** [comment]
The proper fix for this may be to adjust Keyman Developer rather than KeymanWeb, to show the correct default of 15 units for padding. This will then bring keyboards back to reality, and I am considering that for 12.0 stable.

**Comment by mcdurdin** [comment]
Fixed in 12.0.46

**Comment by eddieantonio** [comment]
The updated calculations now:

```
default key width := 100
default key padding := 15

total row width = n * default key width + n * default padding
total row width = 8 * 100 + 8 * 15
total row width = 920
```

```
my key width = k * default key width + (k - 1) * default padding
my key width = 2 * 100 + (2 - 1) * 15
my key width = 215
```

@mcdurdin should this be documented somewhere?

**Comment by mcdurdin** [comment]
> @mcdurdin should this be documented somewhere?

We should document the defaults somewhere. Perhaps in the touch layout editor documentation at https://help.keyman.com/developer/12.0/context/keyboard-editor#toc-touch-layout-tab (that document really needs to be split into multiple pages because it is so long...)
